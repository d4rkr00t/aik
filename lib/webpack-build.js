'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.removeDist = removeDist;

var _fsExtra = require('fs-extra');

var _fsExtra2 = _interopRequireDefault(_fsExtra);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _webpack = require('webpack');

var _webpack2 = _interopRequireDefault(_webpack);

var _gzipSize = require('gzip-size');

var _gzipSize2 = _interopRequireDefault(_gzipSize);

var _configBuilder = require('./webpack/config-builder');

var _configBuilder2 = _interopRequireDefault(_configBuilder);

var _messages = require('./utils/messages');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

/**
 * Removes distribute folder to prevent duplicates.
 */
function removeDist(distPath) {
  return new Promise(resolve => _fsExtra2.default.remove(distPath, resolve));
}

/**
 * Builds project using webpack.
 */

exports.default = (() => {
  var _ref = _asyncToGenerator(function* (filename, flags, params) {
    try {
      _fsExtra2.default.statSync(filename);
    } catch (error) {
      (0, _messages.fileDoesNotExistMsg)(filename);
      return;
    }

    const config = (0, _configBuilder2.default)(filename, flags, params);
    const compiler = (0, _webpack2.default)(config);

    (0, _messages.builderBanner)(filename, flags, params);
    (0, _messages.builderRemovingDistMsg)(params.dist.path);

    yield removeDist(params.dist.path);

    (0, _messages.builderRunningBuildMsg)();

    return new Promise(function (resolve, reject) {
      compiler.run(function (err, stats) {
        if (err) {
          (0, _messages.builderErrorMsg)(err);
          return reject();
        }

        const json = stats.toJson({}, true);
        const buildDuration = stats.endTime - stats.startTime;
        const assets = json.assets.map(function (item) {
          const content = _fsExtra2.default.readFileSync(_path2.default.join(params.dist.path, item.name), 'utf-8');
          return {
            name: item.name,
            size: item.size / 1024,
            sizeGz: _gzipSize2.default.sync(content) / 1024
          };
        });
        (0, _messages.builderSuccessMsg)(params.dist.short, { buildDuration: buildDuration, assets: assets });
        resolve(compiler);
      });
    });
  });

  function runWebpackBuilder(_x, _x2, _x3) {
    return _ref.apply(this, arguments);
  }

  return runWebpackBuilder;
})();