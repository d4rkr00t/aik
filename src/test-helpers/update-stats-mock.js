/* @flow */

import path from "path";
import fs from "fs";
import { execSync } from "child_process";

export const updateStatsMock = function updateStatsMock(name: string) {
  const fixturesDir = path.join(process.cwd(), "__fixtures__", "webpack-stats", name);
  const pathToFixture = path.join(fixturesDir, "index.js");
  const pathToConfig = path.join(fixturesDir, "webpack.config.js");
  const stats = path.join(fixturesDir, "stats.json");
  try {
    execSync(
      `./node_modules/.bin/webpack ${pathToFixture} --output-path='${fixturesDir}' --mode development --config ${pathToConfig} --json > ${stats}`,
      {
        stdio: "ignore"
      }
    );
  } catch (e) {} // eslint-disable-line
  return JSON.parse(fs.readFileSync(stats, "utf8"));
};
