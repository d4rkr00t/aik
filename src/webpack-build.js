/* @flow */

import fs from 'fs-extra';
import path from 'path';
import webpack from 'webpack';
import gzipSize from 'gzip-size';
import webpackConfigBuilder from './webpack/config-builder';
import {
  builderBanner,
  builderRemovingDistMsg,
  builderRunningBuildMsg,
  builderErrorMsg,
  builderSuccessMsg,
  fileDoesNotExistMsg
} from './utils/messages';

/**
 * Removes distribute folder to prevent duplicates.
 */
export function removeDist(distPath: string): Promise<*> {
  return new Promise(resolve => fs.remove(distPath, resolve));
}

/**
 * Builds project using webpack.
 */
export default async function runWebpackBuilder(filename: string, flags: CLIFlags, params: AikParams): Promise<*> {
  try {
    fs.statSync(filename);
  } catch (error) {
    fileDoesNotExistMsg(filename);
    return;
  }

  const config = webpackConfigBuilder(filename, flags, params);
  const compiler = webpack(config);

  builderBanner(filename, flags, params);
  builderRemovingDistMsg(params.dist.path);

  await removeDist(params.dist.path);

  builderRunningBuildMsg();

  return new Promise((resolve, reject) => {
    compiler.run((err, stats) => {
      if (err) {
        builderErrorMsg(err);
        return reject();
      }

      const json = stats.toJson({}, true);
      const buildDuration: number = stats.endTime - stats.startTime;
      const assets:BuildStatAsset[] = json.assets.map(item => {
        const content: string = fs.readFileSync(path.join(params.dist.path, item.name), 'utf-8');
        return {
          name: item.name,
          size: item.size / 1024,
          sizeGz: gzipSize.sync(content) / 1024
        };
      });
      builderSuccessMsg(params.dist.short, { buildDuration, assets });
      resolve(compiler);
    });
  });
}
