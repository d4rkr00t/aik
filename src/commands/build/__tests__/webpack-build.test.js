import path from "path";
import { createBuildStats } from "../webpack-build";
import { updateStatsMock } from "../../../test-helpers/update-stats-mock";

describe("#createBuildStata", () => {
  test("should create build stats from webpack stats json", () => {
    const stats = updateStatsMock("successful-build");
    const buildStats = createBuildStats(
      path.join(process.cwd(), "__fixtures__", "webpack-stats", "successful-build"),
      stats
    );
    expect(typeof buildStats.buildDuration).toBe("number");
    expect(buildStats.assets).toMatchSnapshot();
  });
});
