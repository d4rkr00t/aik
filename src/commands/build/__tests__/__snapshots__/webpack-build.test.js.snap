// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createBuildStata should create build stats from webpack stats json 1`] = `
Array [
  Object {
    "name": "main.js",
    "size": 4.568359375,
    "sizeGz": 1.333984375,
  },
  Object {
    "name": "index.html",
    "size": 0.17578125,
    "sizeGz": 0.146484375,
  },
]
`;
